flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+        extensions: Optional[Tuple[str, ...]] = None,
+    ):
+        super().__init__(
+            train_data=train_folder,
+            val_data=val_folder,
+            test_data=test_folder,
+            predict_data=predict_folder,
+        )
+
+        self.extensions = extensions
+
+    @staticmethod
+    def find_classes(dir: str) -> Tuple[List[str], Dict[str, int]]:
+        """
+        Finds the class folders in a dataset. Ensures that no class is a subdirectory of another.
+
+        Args:
+            dir: Root directory path.
+
+        Returns:
+            tuple: (classes, class_to_idx) where classes are relative to (dir), and class_to_idx is a dictionary.
+        """
+        classes = [d.name for d in os.scandir(dir) if d.is_dir()]
+        classes.sort()
+        class_to_idx = {cls_name: i for i, cls_name in enumerate(classes)}
+        return classes, class_to_idx
+
+    def load_data(self, data: Any, dataset: Optional[Any] = None) -> Iterable[Mapping[str, Any]]:
+        classes, class_to_idx = self.find_classes(data)
+        if not classes:
+            files = [os.path.join(data, file) for file in os.listdir(data)]
+            return [{
+                'input': file
+            } for file in filter(
+                lambda file: has_file_allowed_extension(file, self.extensions),
+                files,
+            )]
+        else:
+            self.preprocess.set_state(LabelsState(classes))
+        dataset.num_classes = len(classes)
+        data = make_dataset(data, class_to_idx, extensions=self.extensions)
+        return [{'input': input, 'target': target} for input, target in data]
+
+
+class FilesDataSource(DataSource, ABC):
+
+    def __init__(
+        self,
+        train_files: Optional[Union[Sequence[Union[str, pathlib.Path]], Iterable[Union[str, pathlib.Path]]]] = None,
+        train_targets: Optional[Union[Sequence[Any], Iterable[Any]]] = None,
+        val_files: Optional[Union[Sequence[Union[str, pathlib.Path]], Iterable[Union[str, pathlib.Path]]]] = None,
+        val_targets: Optional[Union[Sequence[Any], Iterable[Any]]] = None,
+        test_files: Optional[Union[Sequence[Union[str, pathlib.Path]], Iterable[Union[str, pathlib.Path]]]] = None,
+        test_targets: Optional[Union[Sequence[Any], Iterable[Any]]] = None,
+        predict_files: Optional[Union[Sequence[Union[str, pathlib.Path]], Iterable[Union[str, pathlib.Path]]]] = None,
+        extensions: Optional[Tuple[str, ...]] = None,
+    ):
+        super().__init__(
+            train_data=(train_files, train_targets),
+            val_data=(val_files, val_targets),
+            test_data=(test_files, test_targets),
+            predict_data=(predict_files, None),
+        )
+
+        self.extensions = extensions
+
+    def load_data(self, data: Any, dataset: Optional[Any] = None) -> Iterable[Mapping[str, Any]]:
+        # TODO: Bring back the code to work out how many classes there are
```

Comment:
`_load_data_from_tuple` , etc ?

Reply:
Not sure what you mean here, `data` should always be a tuple, just can sometimes be not a tuple if we are predicting.

Reply:
what is the difference with `SequenceDataSource` class ?